toolkit/tools/srpmdownloader/srpmdownloader.go

```diff
@@ -0,0 +1,232 @@
+		if err != nil {
+			return
+		}
+		defer chroot.Close(leaveFilesOnDisk)
+	}
+
+	doCreateAll := func() error {
+		err = getSRPMQuery(specsDir, distTag, buildDir, outDir, workers, runCheck, srpmList, templateSrcConfig)
+		return err
+	}
+
+	if chroot != nil {
+		logger.Log.Info("Grabbing SRPMs URL inside a chroot environment")
+		err = chroot.Run(doCreateAll)
+	} else {
+		logger.Log.Info("Grabbing SRPMs URL SRPMs in the host environment")
+		err = doCreateAll()
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// If this is container build then the bind mounts will not have been created.
+	// Copy the chroot output to host output folder.
+	if !buildpipeline.IsRegularBuild() {
+		srpmsInChroot := filepath.Join(chroot.RootDir(), outDir)
+		err = directory.CopyContents(srpmsInChroot, originalOutDir)
+		if err != nil {
+			return err
+		}
+	}
+
+	return
+}
+
+// getSRPMQuery queries for the name, version and release of the SRPM
+func getSRPMQuery(specsDir, distTag, buildDir, outDir string, workers int, runCheck bool, srpmList []string, srcConfig sourceRetrievalConfiguration) (err error) {
+	const (
+		emptyQueryFormat = ``
+		querySrpm        = `%{NAME}-%{VERSION}-%{RELEASE}.src.rpm`
+	)
+	// Find the general SPEC query arguments
+	defines := rpm.DefaultDefines(runCheck)
+	defines[rpm.DistTagDefine] = distTag
+	arch, err := rpm.GetRpmArch(runtime.GOARCH)
+	if err != nil {
+		logger.Log.Warn(err)
+		return
+	}
+	specFiles, err := srpm.FindSPECFiles(specsDir, srpmList)
+	if err != nil {
+		return
+	}
+
+	// Parse each SPEC for name and version and hydrate SRPM
+	for _, specfile := range specFiles {
+		sourcedir := filepath.Dir(specfile)
+		var packageSRPMs []string
+		packageSRPMs, err = rpm.QuerySPEC(specfile, sourcedir, querySrpm, arch, defines, rpm.QueryHeaderArgument)
+		if err != nil {
+			logger.Log.Warn(err)
+			return err
+		}
+		packageSRPM := packageSRPMs[0]
+		downloadSRPM(packageSRPM, outDir, srcConfig)
+	}
+	return err
+}
+
+// downloadSRPM downloads SRPM trying each of srcConfig's urls
+func downloadSRPM(fileName string, newSourceDir string, srcConfig sourceRetrievalConfiguration) {
+	const (
+		downloadRetryAttempts = 3
+		downloadRetryDuration = time.Second
+	)
+	destinationFile := filepath.Join(newSourceDir, fileName)
+
+	var err error
+	var urlChosen string
+	srpmDownloaded := false
+	for _, urlSrc := range srcConfig.sourceURLs {
+		if srpmDownloaded {
+			break
+		}
+		logger.Log.Debugf("Trying (%s) from (%s)", fileName, urlSrc)
+		url := network.JoinURL(urlSrc, fileName)
+
+		err = retry.Run(func() error {
+			err := network.DownloadFile(url, destinationFile, srcConfig.caCerts, srcConfig.tlsCerts)
+			urlChosen = url
+			if err != nil {
+				logger.Log.Warnf("Failed to download (%s). Error: %s", url, err)
+				os.Remove(destinationFile)
+			} else {
+				srpmDownloaded = true
+			}
+
+			return err
+		}, downloadRetryAttempts, downloadRetryDuration)
+	}
```

Comment:
Slight refactoring **suggestion**, mostly because it immediately triggered suspicion in my head to see a `if srpmDownloaded` check right after `srpmDownloaded := false`.;)
```suggestion
	for _, urlSrc := range srcConfig.sourceURLs {
		logger.Log.Debugf("Trying (%s) from (%s)", fileName, urlSrc)
		url := network.JoinURL(urlSrc, fileName)

		err = retry.Run(func() error {
			err := network.DownloadFile(url, destinationFile, srcConfig.caCerts, srcConfig.tlsCerts)
			urlChosen = url
			srpmDownloaded = (err == nil)
			if !srpmDownloaded {
				logger.Log.Warnf("Failed to download SRPM from URL (%s). Error: %s", url, err)
				os.Remove(destinationFile)
			}

			return err
		}, downloadRetryAttempts, downloadRetryDuration)

		if srpmDownloaded {
			break
		}
	}
```